qiskit/circuit/quantumcircuit.py

```diff
@@ -1737,16 +1740,24 @@ def qasm(
             elif operation.name == "barrier":
                 qargs = ",".join(bit_labels[q] for q in instruction.qubits)
                 instruction_qasm = "barrier;" if not qargs else f"barrier {qargs};"
+            elif hasattr(operation, "has_qasm2") and not operation.has_qasm2():
+                raise CircuitError(
+                    f"No OpenQASM 2 representation is defined for operation {operation.name}"
+                )
             else:
                 operation = _qasm2_define_custom_operation(
                     operation, existing_gate_names, gates_to_define
                 )
-
                 # Insert qasm representation of the original instruction
+                if operation.name == "c3sx":
+                    operation.name = "c3sqrtx"
+                name_param = _instruction_qasm2(operation)
+                if operation.name == "c3sqrtx":
+                    operation.name = "c3sx"
                 bits_qasm = ",".join(
                     bit_labels[j] for j in itertools.chain(instruction.qubits, instruction.clbits)
                 )
-                instruction_qasm = f"{operation.qasm()} {bits_qasm};"
+                instruction_qasm = f"{name_param} {bits_qasm};"
```

Comment:
The ordering of these calls doesn't look right - `_qasm2_define_custom_operation` is what actually defines the operation and it's true recursive.  The special handling should be in there if anything - it already has logic to transform one operation into another for use as the definition and call point.

Reply:
See https://github.com/Qiskit/qiskit-terra/pull/10399#pullrequestreview-1531079795 .